src/devices/Pn532/Pn532.cs

```diff
@@ -0,0 +1,1623 @@
+        private const byte Preamble = 0x00;
+        private const byte Postamble = 0x00;
+        private const byte StartCode1 = 0x00;
+        private const byte StartCode2 = 0xFF;
+        // Operation type for SPI
+        private const byte WriteData = 0b0000_0001;
+        private const byte ReadStatus = 0b0000_0010;
+        private const byte ReadData = 0b0000_0011;
+        // Acknowledge
+        private byte[] AckBuffer = { 0x00, 0x00, 0xFF, 0x00, 0xFF, 0x00 };
+        // Specific buffer to wake up the sensor in serial HSU mode
+        private byte[] SerialWakeUp = new byte[] { 0x55, 0x55, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
+        private byte[] I2cWakeUp = new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 };
+        private int I2cMAxBuffer = 1024;
+        private ParametersFlags _parametersFlags;
+
+        private SpiDevice _spiDevice = null;
+        private I2cDevice _i2cDevice = null;
+        private GpioController _controller = null;
+        private SerialPort _serialPort = null;
+        private int _pin = 18;
+        private SecurityAccessModuleMode _securityAccessModuleMode;
+        private uint _virtualCardTimeout = 20;
+
+        /// <summary>
+        /// The Log level
+        /// </summary>        
+        public LogLevel LogLevel { get { return LogInfo.LogLevel; } set { LogInfo.LogLevel = value; } }
+
+        /// <summary>
+        /// The location to log the info
+        /// </summary>
+        public LogTo LogTo { get { return LogInfo.LogTo; } set { LogInfo.LogTo = value; } }
+
+        /// <summary>
+        /// Firmware version information
+        /// </summary>
+        public FirmwareVersion FirmwareVersion { get; internal set; }
+
+        #region Spi and I2c Settings
+        /// <summary>
+        /// PN532 SPI Clock Frequency
+        /// </summary>
+        public const int SpiClockFrequency = 2_000_000;
+
+        /// <summary>
+        /// Only SPI Mode supported is Mode0
+        /// </summary>
+        public const SpiMode SpiMode = System.Device.Spi.SpiMode.Mode0;
+
+        /// <summary>
+        /// The default I2C address
+        /// </summary>
+        public const byte I2cDefaultAddress = 0x24;
+
+        #endregion        
+
+        /// <summary>
+        /// Create a PN532 using Serial Port
+        /// </summary>
+        /// <param name="portName">The port name</param>
+        public Pn532(string portName)
+        {
+#if DEBUG
+            LogLevel = LogLevel.Debug;
+#endif
+
+            // Data bit : 8 bits,
+            // Parity bit : none,
+            // Stop bit : 1 bit,
+            // Baud rate : 115 200 bauds,
+            LogInfo.Log("Opening serial port 115200, Parity.None, 8 bits, StopBits.One", LogLevel.Debug);
+            _serialPort = new SerialPort(portName, 115200, Parity.None, 8, StopBits.One);
```

Comment:
(optional for the first iteration) it would be useful to create internal abstraction similar to what you've proposed in the issue just for this

Reply:
I was thinking of that. But due to the limitation of the current I2C and SPI, it will be useless, they have to be managed in a different way.

Reply:
what limitations?

Reply:
> what limitations?

Reading and writting are not the same way exactly. Here are the elements:
- for I2c, because we open and close the communication channel every time we read a byte or a byte array, I have to read much more than I should (and it's not possible to know in advance how many bytes will be sent).
- For SPI, there is this issue with the LSB first but that's not what is blocking here, it's just that the protocol is asking for 1 more byte to read, write or check the status.
So for all the 3 way of reading and writting on the sensor, we have to use different logic, even if they are very close to each others.